In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

In [2]:
movies=pd.read_csv("movies_dataset.csv")

In [3]:
movies_df=movies.copy()

In [4]:
movies_df.head()

Unnamed: 0  movie_id                                     title  \
0           0     19995                                    Avatar   
1           1       285  Pirates of the Caribbean: At World's End   
2           2    206647                                   Spectre   
3           3     49026                     The Dark Knight Rises   
4           4     49529                               John Carter   

                                                tags  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...  
3  Following the death of District Attorney Harve...  
4  John Carter is a war-weary, former military ca...

In [5]:
movies_df.drop(["Unnamed: 0"],axis=1,inplace=True)

In [6]:
movies_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...  
3  Following the death of District Attorney Harve...  
4  John Carter is a war-weary, former military ca...

In [7]:
movies_df.isnull().sum()

movie_id    0
title       0
tags        0
dtype: int64

In [8]:
movies_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4806 entries, 0 to 4805
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   movie_id  4806 non-null   int64 
 1   title     4806 non-null   object
 2   tags      4806 non-null   object
dtypes: int64(1), object(2)
memory usage: 112.8+ KB


In [9]:
movies_df["tags"][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. Action Adventure Fantasy ScienceFiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d SamWorthington ZoeSaldana SigourneyWeaver JamesCameron'

In [12]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

In [15]:
tfidf = TfidfVectorizer()
overview_matrix = tfidf.fit_transform(movies_df["tags"])

In [16]:
overview_matrix.shape

(4806, 35831)

In [17]:
count_array=overview_matrix.toarray()

In [18]:
count_array

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [19]:
overview_matrix=pd.DataFrame(data=count_array,columns=tfidf.get_feature_names_out())

In [20]:
overview_matrix

00  000  007  07am   10  100  1000  101  108  10th  ...  única  über  \
0     0.0  0.0  0.0   0.0  0.0  0.0   0.0  0.0  0.0   0.0  ...    0.0   0.0   
1     0.0  0.0  0.0   0.0  0.0  0.0   0.0  0.0  0.0   0.0  ...    0.0   0.0   
2     0.0  0.0  0.0   0.0  0.0  0.0   0.0  0.0  0.0   0.0  ...    0.0   0.0   
3     0.0  0.0  0.0   0.0  0.0  0.0   0.0  0.0  0.0   0.0  ...    0.0   0.0   
4     0.0  0.0  0.0   0.0  0.0  0.0   0.0  0.0  0.0   0.0  ...    0.0   0.0   
...   ...  ...  ...   ...  ...  ...   ...  ...  ...   ...  ...    ...   ...   
4801  0.0  0.0  0.0   0.0  0.0  0.0   0.0  0.0  0.0   0.0  ...    0.0   0.0   
4802  0.0  0.0  0.0   0.0  0.0  0.0   0.0  0.0  0.0   0.0  ...    0.0   0.0   
4803  0.0  0.0  0.0   0.0  0.0  0.0   0.0  0.0  0.0   0.0  ...    0.0   0.0   
4804  0.0  0.0  0.0   0.0  0.0  0.0   0.0  0.0  0.0   0.0  ...    0.0   0.0   
4805  0.0  0.0  0.0   0.0  0.0  0.0   0.0  0.0  0.0   0.0  ...    0.0   0.0   

      đỗthịhảiyến  špelacolja   γη  юлияснигирь  卧底肥妈   张立  绝地奶霸  超级妈妈  
0             0.0         0.0  0.0          0.0   0.0  0.0   0.0   0.0  
1             0.0         0.0  0.0          0.0   0.0  0.0   0.0   0.0  
2             0.0         0.0  0.0          0.0   0.0  0.0   0.0   0.0  
3             0.0         0.0  0.0          0.0   0.0  0.0   0.0   0.0  
4             0.0         0.0  0.0          0.0   0.0  0.0   0.0   0.0  
...           ...         ...  ...          ...   ...  ...   ...   ...  
4801          0.0         0.0  0.0          0.0   0.0  0.0   0.0   0.0  
4802          0.0         0.0  0.0          0.0   0.0  0.0   0.0   0.0  
4803          0.0         0.0  0.0          0.0   0.0  0.0   0.0   0.0  
4804          0.0         0.0  0.0          0.0   0.0  0.0   0.0   0.0  
4805          0.0         0.0  0.0          0.0   0.0  0.0   0.0   0.0  

[4806 rows x 35831 columns]

In [21]:
from sklearn.metrics.pairwise import cosine_similarity

In [22]:
final=pd.DataFrame(cosine_similarity(count_array))

In [23]:
similarity_matrix=final.copy()

In [24]:
similarity_matrix

0         1         2         3         4         5         6     \
0     1.000000  0.028301  0.017183  0.024261  0.079641  0.039612  0.013027   
1     0.028301  1.000000  0.031347  0.021184  0.041685  0.049074  0.016521   
2     0.017183  0.031347  1.000000  0.022886  0.027405  0.028827  0.011407   
3     0.024261  0.021184  0.022886  1.000000  0.025754  0.029991  0.027739   
4     0.079641  0.041685  0.027405  0.025754  1.000000  0.028968  0.041728   
...        ...       ...       ...       ...       ...       ...       ...   
4801  0.015708  0.023869  0.030635  0.032729  0.037639  0.014975  0.024614   
4802  0.010677  0.009761  0.005201  0.010949  0.008840  0.007257  0.008762   
4803  0.020631  0.028020  0.029275  0.045792  0.033013  0.029607  0.036127   
4804  0.013451  0.019850  0.013353  0.046781  0.029701  0.015934  0.029049   
4805  0.007094  0.011713  0.012780  0.026641  0.019233  0.009789  0.025777   

          7         8         9     ...      4796      4797      4798  \
0     0.065816  0.018059  0.018172  ...  0.003616  0.006978  0.035071   
1     0.063842  0.028338  0.033123  ...  0.002105  0.015714  0.030164   
2     0.077295  0.025291  0.021344  ...  0.010343  0.003522  0.013062   
3     0.067242  0.019851  0.151983  ...  0.002605  0.012626  0.023483   
4     0.084898  0.021634  0.036152  ...  0.012671  0.017421  0.031181   
...        ...       ...       ...  ...       ...       ...       ...   
4801  0.060240  0.038213  0.022611  ...  0.007588  0.021340  0.037669   
4802  0.010744  0.003054  0.004524  ...  0.000000  0.006534  0.010743   
4803  0.035608  0.031252  0.025760  ...  0.016584  0.018961  0.043485   
4804  0.030346  0.037406  0.019673  ...  0.002696  0.022759  0.031886   
4805  0.022554  0.026550  0.020820  ...  0.002922  0.007308  0.054200   

          4799      4800      4801      4802      4803      4804      4805  
0     0.034464  0.014465  0.015708  0.010677  0.020631  0.013451  0.007094  
1     0.034941  0.043802  0.023869  0.009761  0.028020  0.019850  0.011713  
2     0.013786  0.029785  0.030635  0.005201  0.029275  0.013353  0.012780  
3     0.025531  0.031535  0.032729  0.010949  0.045792  0.046781  0.026641  
4     0.027378  0.044481  0.037639  0.008840  0.033013  0.029701  0.019233  
...        ...       ...       ...       ...       ...       ...       ...  
4801  0.019804  0.056922  1.000000  0.003609  0.021091  0.041227  0.026576  
4802  0.015006  0.008643  0.003609  1.000000  0.025532  0.016957  0.005606  
4803  0.038865  0.025103  0.021091  0.025532  1.000000  0.041835  0.030052  
4804  0.031459  0.037660  0.041227  0.016957  0.041835  1.000000  0.038382  
4805  0.015931  0.031843  0.026576  0.005606  0.030052  0.038382  1.000000  

[4806 rows x 4806 columns]

In [30]:
similarity[4] = similarity_matrix[4]

In [31]:
similarity_score = list(enumerate(similarity[4]))

In [32]:
similarity_score

[(0, 0.07964054910883143),
 (1, 0.04168462357190541),
 (2, 0.027404766916246704),
 (3, 0.025754136563104425),
 (4, 1.0000000000000002),
 (5, 0.028967571527133244),
 (6, 0.04172809249569412),
 (7, 0.08489779362712627),
 (8, 0.021633881234195143),
 (9, 0.036151945714320614),
 (10, 0.07253007463261378),
 (11, 0.03525087016499159),
 (12, 0.04647271536625517),
 (13, 0.03144872028586676),
 (14, 0.05527342228690751),
 (15, 0.04203673620696268),
 (16, 0.06537462935119774),
 (17, 0.07218359975927946),
 (18, 0.061485897498032356),
 (19, 0.052163452814337825),
 (20, 0.05859804376783131),
 (21, 0.049117046856501374),
 (22, 0.017725548918984424),
 (23, 0.023471985906501607),
 (24, 0.033045618802632806),
 (25, 0.046351462716147165),
 (26, 0.08206831390350813),
 (27, 0.09373720788145946),
 (28, 0.044888981453342504),
 (29, 0.04055855277496957),
 (30, 0.030380260025403807),
 (31, 0.04093651305361655),
 (32, 0.06445389776923216),
 (33, 0.03835061301803708),
 (34, 0.0143389573030968),
 (35, 0.0381771801

In [33]:
similarity_score=sorted(enumerate(similarity[4]),key=lambda x:x[1],reverse=True)

In [34]:
similarity_score

[(4, 1.0000000000000002),
 (373, 0.1652768721699361),
 (1741, 0.14559137881647713),
 (1257, 0.14440964937138862),
 (270, 0.13110962683660954),
 (487, 0.13107406310064065),
 (473, 0.12743628671300958),
 (4166, 0.12496883670901456),
 (2968, 0.11475524966840567),
 (207, 0.11329724203436174),
 (2760, 0.10939153809238279),
 (127, 0.10819833168300652),
 (752, 0.10554616145708542),
 (85, 0.10306099909063716),
 (2936, 0.10140847974049345),
 (3377, 0.1005556548146428),
 (1322, 0.09810382742293162),
 (27, 0.09373720788145946),
 (345, 0.09344974556000389),
 (870, 0.09230863291152946),
 (47, 0.09207872826525239),
 (833, 0.09205479173652503),
 (581, 0.09190702946195735),
 (322, 0.08935751755540279),
 (4656, 0.08908913002710621),
 (778, 0.08811184680255295),
 (3910, 0.08783641904964126),
 (48, 0.08645476919669048),
 (7, 0.08489779362712627),
 (1310, 0.08486529118404063),
 (91, 0.08389464465494088),
 (1965, 0.08364962739202839),
 (1161, 0.0835716135143523),
 (230, 0.08349973360591958),
 (610, 0.08293

in similarity_matrix 0 to 4806 are the movies and their similar movies.

It follows the below steps to make recommendations.

Get the index of the movie using the title. Get the list of similarity scores of the movies concerning all the movies. Enumerate them (create tuples) with the first element being the index and the second element is the cosine similarity score. Sort the list of tuples in descending order based on the similarity score. Get the list of the indices of the top 10 movies from the above sorted list. Exclude the first element because it is the title itself. Map those indices to their respective titles and return the movies list.

Now, let us create a series that maps the index of the matrix to movie names to make it easy for us to just feed in movie names and get the recommendation.

In [35]:
#movie_index_mapping
mapping=pd.Series(movies_df.index,index=movies_df["title"])
mapping

title
Avatar                                         0
Pirates of the Caribbean: At World's End       1
Spectre                                        2
The Dark Knight Rises                          3
John Carter                                    4
                                            ... 
El Mariachi                                 4801
Newlyweds                                   4802
Signed, Sealed, Delivered                   4803
Shanghai Calling                            4804
My Date with Drew                           4805
Length: 4806, dtype: int64

In [36]:
movie_index=mapping["John Carter"]
movie_index

4

In [37]:
# create a function that will take the movie name and make recommendation

def recommend_movie(movie_input):
    movie_index = mapping[movie_input]
    # get similarity value with other movies
    similarity_score = list(enumerate(similarity[movie_index]))
    similarity_score = sorted(similarity_score, key=lambda x: x[1], reverse=True)
    # get the list of top 10 recommended movies
    similarity_score = similarity_score[:10]

    movie_ind = [i[0] for i in similarity_score]
    recommended_movies = movies['title'].iloc[movie_ind]
    print('The recommended movies:', movie_input,"is:")
    return recommended_movies

In [38]:
recommend_movie("John Carter")

The recommended movies: John Carter is:


4                       John Carter
373                 Mission to Mars
1741                 Ghosts of Mars
1257                     Get Carter
270                     The Martian
487                      Red Planet
473                   Mars Attacks!
4166    The Marine 4: Moving Target
2968          The Last Days on Mars
207                    Total Recall
Name: title, dtype: object

In [39]:
recommend_movie("Total Recall")

The recommended movies: Total Recall is:


207              Total Recall
1025            The Cable Guy
487                Red Planet
1741           Ghosts of Mars
270               The Martian
373           Mission to Mars
2968    The Last Days on Mars
4                 John Carter
473             Mars Attacks!
4129           Roadside Romeo
Name: title, dtype: object